### PR TITLE
[zh-cn] update setTimeout API

### PR DESCRIPTION
**Pull request by jasonren0403**
### Related issues and pull requests

follow-up of #13009, closed by mistake

### CODE REVIEW

File: files/zh-cn/web/api/settimeout/index.md [hunk 1]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
-## 关于"`this`"的问题
-
-当你向 `setTimeout()` (或者其他函数) 传递一个函数时，该函数中的`this`指向跟你的期望可能不同，这个问题在 [JavaScript reference](/zh-CN/docs/JavaScript/Reference/Operators/this#Method_binding) 中进行了详细解释。
-
-### 解释
-
-由`setTimeout()`调用的代码运行在与所在函数完全分离的执行环境上。这会导致，这些代码中包含的 `this` 关键字在非严格模式会指向 `window` (或全局) 对象，严格模式下为 undefined，这和所期望的`this`的值是不一样的。
-
-> **备注：** 即使是在严格模式下，`setTimeout()`的回调函数里面的`this`仍然默认指向`window`对象，并不是`undefined`。
+当你向 `setTimeout()` 传递一个函数时，该函数中的 `this` 指向跟你的期望可能不同，这个问题在 [JavaScript 参考](/zh-CN/docs/JavaScript/Reference/Operators/this#回调)中进行了详细解释。
```
Comment (jasonren0403) [inline files/zh-cn/web/api/settimeout/index.md:109]:
```suggestion
当你向 `setTimeout()` 传递一个函数时，该函数中的 `this` 指向跟你的期望可能不同，这个问题在 [JavaScript 参考](/zh-CN/docs/Web/JavaScript/Reference/Operators/this#回调)中进行了详细解释。
```

File: files/zh-cn/web/api/settimeout/index.md [hunk 2]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
+```
 
-> **备注：** 4 ms 是在 [HTML5 spec](http://www.whatwg.org/specs/web-apps/current-work/multipage/timers.html#timers) 中精确的，并且在 2010 年及以后的跨浏览器中保持了一致，这个数值比 Gecko 5.0 规定的嵌套函数的最小延时 10ms 更为精确。
+出现这个结果的原因是，尽管 `setTimeout` 以 0ms 的延迟来调用函数，但这个任务已经被放入了队列中并且等待下一次执行；并不是立即执行；队列中的等待函数被调用之前，当前代码必须全部运行完毕，因此这里运行结果并非预想的那样。
 
-#### 未被激活的 tabs 的定时最小延迟>=1000ms
+### 在加载页面时推迟超时
 
-为了优化后台 tab 的加载损耗（以及降低耗电量），在未被激活的 tab 中定时器的最小延时限制为 1S(1000ms)。
+当前标签页正在加载时，Firefox 将推迟触发 `setTimeout()` 计时器。直到主线程被认为是空闲的（类似于 [window.requestIdleCallback()](/zh-CN/docs/Web/API/window/requestIdleCallback)），或者直到加载事件触发完毕，才开始触发。
```
Comment (jasonren0403) [inline files/zh-cn/web/api/settimeout/index.md:319]:
```suggestion
当前标签页正在加载时，Firefox 将推迟触发 `setTimeout()` 计时器。直到主线程被认为是空闲的（类似于 [window.requestIdleCallback()](/zh-CN/docs/Web/API/Window/requestIdleCallback)），或者直到加载事件触发完毕，才开始触发。
```

File: files/zh-cn/web/api/settimeout/index.md [hunk 3]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
 
-function delayedAlert() {
-  timeoutID = window.setTimeout(slowAlert, 2000);
-}
+### 非数字延迟值被静默地强制转化为数字
 
-function slowAlert() {
-  alert('That was really slow!');
-}
+如果调用 `setTimeout()` 的 [_delay_](#delay) 值不是数字，隐含的[类型强制转换](/zh-CN/docs/Glossary/Type_coercion)会默默地对该值进行转换，使其成为数字。例如，下面的代码在 _delay_ 值中错误地使用了字符串 `"1000"`，而不是数字 `1000`——但它仍然有效，因为当代码运行时，字符串被强制转换成数字 `1000`，所以代码在 1 秒后执行。
```
Comment (yin1999) [inline files/zh-cn/web/api/settimeout/index.md:55]:
```suggestion
如果调用 `setTimeout()` 的 [_delay_](#delay) 值不是数字，隐含的[类型强制转换](/zh-CN/docs/Glossary/Type_coercion)会静默地对该值进行转换，使其成为数字。例如，下面的代码在 _delay_ 值中错误地使用了字符串 `"1000"`，而不是数字 `1000`——但它仍然有效，因为当代码运行时，字符串被强制转换成数字 `1000`，所以代码在 1 秒后执行。
```

File: files/zh-cn/web/api/settimeout/index.md [hunk 4]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
-> **备注：** 这两个替换也让 IE 支持了符合 HTML5 标准的定时器函数。所以也能作为一个 polyfills。查看 [Callback arguments](#Callback_arguments) 一段。
+```js example-good
+// 这样做
+setTimeout(() => {
+  console.log("Hello World!");
+}, 500);
+```
 
-新特性检测：
+传递给 {{domxref("setTimeout()")}} 的字符串是在全局上下文中求值的，因此当字符串被求值为代码时，{{domxref("setTimeout()")}} 被调用的上下文中的局部符号将不可用。
```
Comment (yin1999) [inline files/zh-cn/web/api/settimeout/index.md:197]: mdn/content#26512

```suggestion
传递给 `setTimeout()` 的字符串是在全局上下文中求值的，因此当字符串被求值为代码时，`setTimeout()` 被调用的上下文中的局部符号将不可用。
```

File: files/zh-cn/web/api/settimeout/index.md [hunk 5]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
+### 延时比指定值更长的原因
 
-setTimeout(alert, 1500, "Hello world!"); // the standard use of setTimeout and setInterval is preserved, but...
-setTimeout.call(myArray, myArray.myMethod, 2000); // prints "zero,one,two" after 2 seconds
-setTimeout.call(myArray, myArray.myMethod, 2500, 2); // prints "two" after 2,5 seconds
-```
+有很多因素会导致 setTimeout 的回调函数执行比设定的预期值更久，本节将讨论最常见的原因。
 
-针对这个问题并没有原生的解决方案。
+#### 嵌套延时
```
Comment (yin1999) [inline files/zh-cn/web/api/settimeout/index.md:203]:
```suggestion
#### 嵌套超时
```

File: files/zh-cn/web/api/settimeout/index.md [hunk 6]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
 
-向`setTimeout()`传递一个字符串而不是函数会遭受到与使用[`eval`](/zh-CN/docs/JavaScript/Reference/Global_Objects/eval#Don.27t_use_eval.21)一样的风险。
+{{EmbedLiveSample("嵌套延时", 100, 420)}}
 
-```js
-// 推荐
-window.setTimeout(function() {
-    alert("Hello World!");
-}, 500);
+#### 标签的延时
```
Comment (yin1999) [inline files/zh-cn/web/api/settimeout/index.md:280]:
```suggestion
#### 非活动标签的超时
```

File: files/zh-cn/web/api/settimeout/index.md [hunk 7]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
+### WebExtension 背景页面和定时器
 
-Android 版的 Firefox 对未被激活的后台 tabs 的使用了 15min 的最小延迟间隔时间，并且这些 tabs 也能完全不被加载。
+在 [WebExtension](/zh-CN/docs/Mozilla/Add-ons/WebExtensions) 中，`setTimeout()` 不会可靠工作。扩展开发者应当使用 [`alarms`](/zh-CN/docs/Mozilla/Add-ons/WebExtensions/API/alarms) API 作为替代。
 
-> **备注：** 当 Web Audio API {{domxref("AudioContext")}} 正在被用来播放音频的时候，Firefox 50 不会再限制后台 tabs 的加载。后续的 Firefox 51 版本修正了这个问题，即使在没有音频播放的情况下，也不再限制后台 tabs 的加载。这个解决了一些软件应用在后台 tabs 中播放基于文本的音频 ( note-based) 时，无法去同步音频和画面的问题。
+### 最大延时值
 
-#### 追踪型脚本的最小延时限制
+浏览器内部以 32 位带符号整数存储延时。这就会导致如果一个延时 大于 2147483647 毫秒（大约 24.8 天）时就会溢出，导致定时器将会被立即执行。
```
Comment (jasonren0403) [inline files/zh-cn/web/api/settimeout/index.md:327]:
```suggestion
浏览器内部以 32 位带符号整数存储延时。这就会导致如果一个延时大于 2147483647 毫秒（大约 24.8 天）时就会溢出，导致定时器将会被立即执行。
```

File: files/zh-cn/web/api/settimeout/index.md [hunk 8]
```diff
@@ -9,381 +9,368 @@ slug: Web/API/setTimeout
...
+const myArray = ["zero", "one", "two"];
 myArray.myMethod = function (sProperty) {
-    alert(arguments.length > 0 ? this[sProperty] : this);
+  console.log(arguments.length > 0 ? this[sProperty] : this);
 };
 
-myArray.myMethod(); // prints "zero,one,two"
-myArray.myMethod(1); // prints "one"
+myArray.myMethod(); // 打印 "zero,one,two"
+myArray.myMethod(1); // 打印 "one"
```
Comment (jasonren0403) [inline files/zh-cn/web/api/settimeout/index.md:120]:
```suggestion
myArray.myMethod(); // 输出 "zero,one,two"
myArray.myMethod(1); // 输出 "one"
```

**Review by yin1999 — APPROVED** [review]

**Pull request merged**